add timeout for exec interface

zhouhaibing089 opened the pull request:
This should get us away from situations like https://github.com/kubernetes/kubernetes/issues/63331.

A little bit more context, the `os/exec` package starts to accept `context.Context` in golang 1.7. We should leverage that so we can have a more predictable behavior, then. 

```release-note
NONE
```

zhouhaibing089 committed:
iptables: add timeout when checking rules

in cases where iptables stucks forever due to some reasons, we lost
the availability of kube-proxy, this is about adding a timeout for
the rule checking operations, as a result, it should give us a more
reliable working iptables proxy.

dims commented on pkg/util/iptables/iptables.go:
```diff
@@ -413,11 +415,18 @@ func iptablesCommand(protocol Protocol) string {
+	if ctx == nil {
+		return runner.exec.Command(iptablesCmd, fullArgs...).CombinedOutput()
+	}
+	return runner.exec.CommandContext(ctx, iptablesCmd, fullArgs...).CombinedOutput()
```
Should we not check if we timed out? `ctx.Err() == context.DeadlineExceeded`

zhouhaibing089 replied:
@dims: We should, but in the place where the context object is constructed, see L504 in the same file.

dims replied:
Ah thanks! @zhouhaibing089

dims commented:
@zhouhaibing089 do we need to fix the code here first? https://github.com/kubernetes/utils/tree/master/exec (and then vendor it back in?)

zhouhaibing089 commented:
@dims: Oh, I thought `k8s.io/utils` is another staging repo. Will do!

zhouhaibing089 commented:
Well, so https://github.com/kubernetes/utils/pull/27 already have this.

dims commented:
@zhouhaibing089 ah cool, let's update vendorized to latest code from that repo (and update Godeps.json etc...)

zhouhaibing089 committed:
dep: upgrade k8s.io/utils

update the dependency of k8s.io/utils so we can have CommandContext
in exec package.

zhouhaibing089 commented:
/assign @cblecker

dims commented on Godeps/Godeps.json:
```diff
@@ -2363,82 +2363,82 @@
 		{
 			"ImportPath": "github.com/opencontainers/runc/libcontainer",
-			"Comment": "v1.0.0-rc5-46-g871ba2e5",
+			"Comment": "v1.0.0-rc5-46-g871ba2e",
```
we did not update this component, can you please change the SHA back to original? Looks like your environment chopped off the last digit

zhouhaibing089 replied:
@dims: I am not sure which part I did wrong, I followed the steps at https://github.com/kubernetes/community/blob/master/contributors/devel/godep.md which is pretty helpful.

1.  ./hack/run-in-gopath.sh godep restore
1.  ./hack/run-in-gopath.sh go get k8s.io/utils/exec
1.  ./hack/run-in-gopath.sh godep save

The stripping of last digt might be related with which part of my environment, git version?

cblecker replied:
Yes, please update your git version.

dims commented:
@zhouhaibing089 lot of BUILD files were removed ... can you please add them back?

dims commented:
/ok-to-test

zhouhaibing089 committed:
run ./hack/update-bazel.sh

the previous run of godep save removes the whole BUILD files inside
vendor directory, I have to run ./hack/update-bazel.sh to get them
back.

zhouhaibing089 committed:
dep: run godep save again

thockin commented:
/lgtm
/approve
